### PR TITLE
voting refactor, tests, minor lint changes

### LINKED ISSUE #40: Refactor /articles/votes and add tests
Currently the keywords are being modified in the article.js, and should be in keyword.js. Also there aren't any tests for this yet.

What the tests should be for:
`positive vote with post body { id, vote: 1 }`
`negative vote with post body { id, vote: -1 }`
`invalid id with post body { fakeId, vote: 1 }`
`invalid vote with post body { id, vote: 'badViote' }`

And all of those again but testing that the keywords are modified properly or not.

### PR DESCRIPTION
**Pull request by kirinrastogi**
## Voting
- Now the modification of the keywords is done in `keyword.js`
- Keyword modification now uses `Promise.all`
- Cleaned up `incrementKeyword` to only use .catch to handle errors
- The tests expect the response to have an array of keyword objects, each with `{ word, newSum, oldSum, id}`
- That last point is important, see TODO

## Lint changes
- allow param reassign in `routes/`, I think it was not seeing that Mongoose was modifying the parameters
- remove max len, unused vars, and unused expressions because initializing `should` was tripping this, and some of Jacky's tests

## Tests for posting a vote
- Article gets modified properly
- Keywords get modified properly
- Error handling

## Related TODOs
- Change articles to have an array of keyword id's instead of array of strings #56
- `GET /keywords/:id` should be an endpoint, instead of `GET /keywords/:word` #57

Fixes #40

### CODE REVIEW

**Comment by JackyChiu** [comment]
Looks like you could do quite a bit a squashing,  like ex. your first 3 commits can definitely be squashed together due to the same concerns. 
```
Add test and helper functions for main use	
Test cleanup, and /vote 
Test for negative vote
```

File: routes/article.js [hunk 1]
```diff
@@ -1,6 +1,6 @@
 const express = require('express')
-const Article = require('../models/ArticleModel')
-const Keyword = require('../models/KeywordModel')
+const Article = require('../models/ArticleModel.js')
```
Comment (VikramBombhi) [inline routes/article.js]: Why did you add the .js extention here? I thought it will assume .js if none is given which to me looked better

File: routes/keyword.js [hunk 2]
```diff
@@ -1,53 +1,81 @@
...
+                const len = keyword.votes.length
+                oldSum = keyword.votes[len - 1].sum
+              }
+              const newSum = oldSum + amount
+              const newVote = { sum: newSum, time: Date.now() }
+              keyword.votes.push(newVote)
+              keyword.save()
+              resolve({ word: keyword.word, newSum, oldSum, id: keyword.id })
+            })
+            .catch(err => new Error(err.message))
```
Comment (VikramBombhi) [inline routes/keyword.js]: Fix tabbing please. The whole section is indented alot, is it because ur trying to line up the .findOne and the .then? because i noticed u did this style of tabbing elsewhere as well

File: test/article.test.js [hunk 3]
```diff
@@ -61,168 +102,223 @@ describe('/articles Route', () => {
...
+    const id = '5846399f9ba87af2501fb035'
+    const fakeId = '5846399f9ba87af2501fb03f'
+    const test = getArticleData(id)
+
+    it('should respond with the changed information when a valid positive vote is supplied', (done) => {
+      test.then(articleResponse => articleResponse.body)
+                .then(oldArticle => postVote(id, 1, oldArticle))
+                .then((voteResponse) => {
+                  checkVoteResponse(voteResponse.res)
+                  checkKeywords(voteResponse.res.body.keywords, 1)
```
Comment (VikramBombhi) [inline test/article.test.js]: Can we change the name of checkKeywords()? not a big deal but it should reflect the fact that its checking the response like the checkVoteResponse() so if i read the function name alone I know what it concerns

File: test/article.test.js [hunk 4]
```diff
@@ -61,168 +102,223 @@ describe('/articles Route', () => {
...
+    })
+
+    it('should respond with an error when an invalid vote is supplied', (done) => {
+      test.then(() => postVote(id, 'badVote', 'error'))
+                .then((errorResponse) => {
+                  checkValidError(errorResponse.res)
+                  errorResponse.res.body.givens.id.should.equal(id)
+                  done()
+                })
+                .catch(err => done(err))
```
Comment (VikramBombhi) [inline test/article.test.js]: Good job with all these tests they look pretty good to me

File: routes/article.js (hunk 1 above)
Comment (kirinrastogi) [inline routes/article.js]: I have no idea why I did this.

File: routes/keyword.js (hunk 2 above)
Comment (kirinrastogi) [inline routes/keyword.js]: This is weird, eslint doesn't change this unless I manually set it to be indented a tiny bit.

Comment (kirinrastogi) [inline routes/keyword.js]: @JackyChiu @VikramBombhi @Tuanhao What do we want to do about this? The linter isnt catching is so far.

**Comment by JackyChiu** [comment]
Nice the commits look good 😄

File: routes/keyword.js (hunk 2 above)
Comment (JackyChiu) [inline routes/keyword.js]: When you use eslint to fix when theres tabbing of 4 it gets confused, after we fix everything it's happy

File: test/article.test.js [hunk 5]
```diff
@@ -2,16 +2,51 @@ const chai = require('chai')
...
 
-chai.should()
-chai.use(chaiHttp)
+const should = chai.should()
 
-function checkResponse(res) {
+function postVote(id, vote, old) {
+  return new Promise((resolve) => {
+    chai.request(app)
+            .post('/articles/vote')
```
Comment (JackyChiu) [inline test/article.test.js]: Tabbing is also off here

File: test/article.test.js [hunk 6]
```diff
@@ -2,16 +2,51 @@ const chai = require('chai')
...
+
+function checkReponse(res) {
   res.should.have.status(200)
   res.body.should.be.an('array')
   res.body.should.not.be.empty
   res.body.length.should.be.at.most(10)
 }
 
+function checkValidError(res) {
+  res.should.have.status(500)
```
Comment (JackyChiu) [inline test/article.test.js]: We should check if it's not 200 just incase we decide to change the error statuses sent out, I think they should be in the 400's

File: test/article.test.js [hunk 7]
```diff
@@ -30,11 +65,17 @@ function checkArticleProperties(res) {
...
-  res.body.should.have.property('err')
-  res.body.should.have.property('givens')
+function checkKeyword(keyword, amount) {
+  keyword.should.have.property('oldSum').be.a('number')
+  keyword.should.have.property('newSum').be.a('number')
+  keyword.newSum.should.equal(keyword.oldSum + amount)
+}
+
+function checkKeywords(keywords, amount) {
+  keywords.length.should.be.above(0)
```
Comment (JackyChiu) [inline test/article.test.js]: I think `keywords.should.not.be.empty` is more clear

File: test/article.test.js [hunk 8]
```diff
@@ -61,168 +102,223 @@ describe('/articles Route', () => {
   describe('GET /articles/top', () => {
     it('should have right format response and right properties in articles', (done) => {
       chai.request(app)
-        .get('/articles/top')
-        .end((err, res) => {
-          checkResponse(res)
-          checkArticleProperties(res)
-          done()
-        })
+                .get('/articles/top')
```
Comment (JackyChiu) [inline test/article.test.js]: Tabbing is off here too

File: test/article.test.js [hunk 9]
```diff
@@ -45,11 +86,11 @@ function checkVotesSorted(res) {
 
 function checkDateSorted(res) {
   const articles = res.body
-  const sorted = articles
-    .every((val, i, arr) => i === 0 || arr[i - 1].publishedAt >= val.publishedAt)
+  const sorted = articles.every((val, i, arr) => i === 0 || arr[i - 1].publishedAt >= val.publishedAt)
   sorted.should.be.true
 }
 
+chai.use(chaiHttp)
```
Comment (JackyChiu) [inline test/article.test.js]: Should keep this at the top of the file

File: test/article.test.js [hunk 10]
```diff
@@ -2,16 +2,51 @@ const chai = require('chai')
...
+    chai.request(app)
+            .get(`/articles/${id}`)
+            .end((err, res) => {
+              if (err) reject(err)
+              resolve(res)
+            })
+  })
+}
+
+function checkReponse(res) {
```
Comment (JackyChiu) [inline test/article.test.js]: You removed @VikramBombhi's fix of the correct spelling for the function

File: test/article.test.js [hunk 11]
```diff
@@ -2,16 +2,51 @@ const chai = require('chai')
 const chaiHttp = require('chai-http')
 const app = require('../server.js')
 
-chai.should()
```
Comment (JackyChiu) [inline test/article.test.js]: Why remove?

**Comment by JackyChiu** [comment]
> Change articles to have an array of keyword id's instead of array of strings #56
I'd purpose this to be better for the array
```
[
  word: 'ConUHacks BB',

File: routes/.eslintrc.json [hunk 12]
```diff
@@ -0,0 +1,6 @@
+{
+  "rules": {
+    "no-param-reassign": "off"
```
Comment (JackyChiu) [inline routes/.eslintrc.json]: Why turn this off?

File: routes/article.js [hunk 13]
```diff
@@ -21,53 +21,37 @@ function getAmount(vote) {
...
-              votes: [{
-                sum: amount,
-                time: Date.now(),
-              }],
-            })
-            keyword.save()
+        .then((article) => {
+          if (!article) throw new Error('Article not found')
+          article.votes += amount
+          return article.save()
```
Comment (JackyChiu) [inline routes/article.js]: A better approach would be to use `findByIdAndUpdate`, so there would only be one request sent to the database server inside of querying the database, getting and modifying the response and then requesting to save. Also then no modifications to params 😎 
```js
Article.findByIdAndUpdate(
  req.body.id,
  { $inc : { votes: amount } },
  {
    new: true,
    runValidators: true
   })
   .then((article) => ...
```

Heres the docs : http://mongoosejs.com/docs/api.html#model_Model.findByIdAndUpdate

File: routes/keyword.js [hunk 14]
```diff
@@ -1,53 +1,81 @@
 const express = require('express')
-const Keyword = require('../models/KeywordModel')
+const Keyword = require('../models/KeywordModel.js')
```
Comment (JackyChiu) [inline routes/keyword.js]: You added `.js` here too

File: test/.eslintrc.json [hunk 15]
```diff
@@ -1,5 +1,7 @@
 {
   "rules": {
-    "no-unused-expressions": "off"
+    "no-unused-expressions": "off",
+    "max-len": "off",
+    "no-unused-vars": "off"
```
Comment (JackyChiu) [inline test/.eslintrc.json]: > "no-unused-vars": "off"

Why?

File: routes/keyword.js [hunk 16]
```diff
@@ -1,53 +1,81 @@
...
 })
 
-module.exports = router
+function incrementKeyword(word, amount) {
+  return new Promise((resolve) => {
+    Keyword.findOne({ word })
+            .then((keyword) => {
+              let oldSum = 0
+              if (!keyword) {
+                keyword = new Keyword({
```
Comment (JackyChiu) [inline routes/keyword.js]: This looks out of place, let's move the logic of adding a new keyword to the same time as they are created with the article

Comment (JackyChiu) [inline routes/keyword.js]: (This will also help with the creation of the keywords array in the article schema)

File: routes/keyword.js [hunk 17]
```diff
@@ -1,53 +1,81 @@
...
+                  word,
+                  votes: [],
+                })
+              } else {
+                const len = keyword.votes.length
+                oldSum = keyword.votes[len - 1].sum
+              }
+              const newSum = oldSum + amount
```
Comment (JackyChiu) [inline routes/keyword.js]: Let's refactor the schema to contain a `sum` in the keyword bc the method of getting `oldSum` here is not ideal, this logic will then change ^

File: test/article.test.js [hunk 18]
```diff
@@ -2,16 +2,51 @@ const chai = require('chai')
...
+            })
+  })
+}
+
+function getArticleData(id) {
+  return new Promise((resolve, reject) => {
+    chai.request(app)
+            .get(`/articles/${id}`)
+            .end((err, res) => {
+              if (err) reject(err)
```
Comment (JackyChiu) [inline test/article.test.js]: Should be consistent with checking for error, the above function doesn't, you can pick to check or not

File: test/article.test.js [hunk 19]
```diff
@@ -61,168 +102,223 @@ describe('/articles Route', () => {
...
+                  res.body.should.be.array
+                  res.body.should.be.empty
+                  done()
+                })
+    })
+  })
+
+  describe('POST /articles/vote', () => {
+    const id = '5846399f9ba87af2501fb035'
+    const fakeId = '5846399f9ba87af2501fb03f'
```
Comment (JackyChiu) [inline test/article.test.js]: I think maybe we should put these in a higher scope describe, since they're being used more than once

File: test/article.test.js [hunk 20]
```diff
@@ -61,168 +102,223 @@ describe('/articles Route', () => {
...
+    const test = getArticleData(id)
```
Comment (JackyChiu) [inline test/article.test.js]: I think you should have a better name than test

File: test/article.test.js [hunk 21]
```diff
@@ -61,168 +102,223 @@ describe('/articles Route', () => {
...
+
+    it('should respond with the changed information when a valid positive vote is supplied', (done) => {
+      test.then(articleResponse => articleResponse.body)
```
Comment (JackyChiu) [inline test/article.test.js]: test returns a promise you it's okay to do 
```js
test
  .then(...
```

File: test/article.test.js [hunk 22]
```diff
@@ -2,16 +2,51 @@ const chai = require('chai')
...
+const should = chai.should()
 
-function checkResponse(res) {
+function postVote(id, vote, old) {
+  return new Promise((resolve) => {
+    chai.request(app)
+            .post('/articles/vote')
+            .send({ id, vote })
+            .end((err, res) => {
+              resolve({ res, old })
```
Comment (JackyChiu) [inline test/article.test.js]: This is quite a hacky way to pass down `old` in the promise, make a variable for it instead

File: test/article.test.js [hunk 23]
```diff
@@ -61,168 +102,223 @@ describe('/articles Route', () => {
...
+                .end((err, res) => {
+                  checkReponse(res)
+                  checkArticleProperties(res)
+                  done()
+                })
     })
 
     it('should be in sorted order from greatest to least', (done) => {
+      const lastDate = new Date()
+      lastDate.setDate(lastDate.getDate() - 1)
```
Comment (JackyChiu) [inline test/article.test.js]: Why are these getting set in here? Shouldn't they be in higher scope describe?

File: test/article.test.js (hunk 21 above)
Comment (JackyChiu) [inline test/article.test.js]: Why make a `then` block for returning an always since value?

File: routes/article.js [hunk 24]
```diff
@@ -21,53 +21,37 @@ function getAmount(vote) {
...
+        .then((data) => {
+          Promise.all(incrementKeywords(data.keywords, amount))
+                .then((keywords) => {
+                  res.send({
+                    votes: data.votes,
+                    id: data.id,
+                    keywords,
+                  })
+                })
+                .catch(err => new Error(err.message))
```
Comment (JackyChiu) [inline routes/article.js]: Why not 
```js
 .catch(err => err)
```

File: routes/article.js [hunk 25]
```diff
@@ -21,53 +21,37 @@ function getAmount(vote) {
...
-                time: Date.now(),
-              }],
-            })
-            keyword.save()
+        .then((article) => {
+          if (!article) throw new Error('Article not found')
+          article.votes += amount
+          return article.save()
+        })
+        .then((data) => {
```
Comment (JackyChiu) [inline routes/article.js]: It's not clear what `data` is here, I think you should use`article` instead

File: routes/article.js [hunk 26]
```diff
@@ -1,6 +1,6 @@
 const express = require('express')
-const Article = require('../models/ArticleModel')
-const Keyword = require('../models/KeywordModel')
+const Article = require('../models/ArticleModel.js')
+const incrementKeywords = require('./keyword.js').incrementKeywords
```
Comment (JackyChiu) [inline routes/article.js]: ✂️ `.js` here

File: routes/keyword.js (hunk 16 above)
Comment (JackyChiu) [inline routes/keyword.js]: Lets move this to a new PR

File: routes/keyword.js (hunk 17 above)
Comment (JackyChiu) [inline routes/keyword.js]: Keep for new PR

File: routes/keyword.js [hunk 27]
```diff
@@ -1,53 +1,81 @@
...
+            return accumulator
+          }, {
+            sum: 0,
+            words: [],
+          })
+          res.send(topWords.words)
+        })
+        .catch((err) => {
+          res.status(500).send({
+            err,
```
Comment (JackyChiu) [inline routes/keyword.js]: We should have `err.toString()` here

File: routes/keyword.js [hunk 28]
```diff
@@ -1,53 +1,81 @@
...
+              keyword.votes.push(newVote)
+              keyword.save()
+              resolve({ word: keyword.word, newSum, oldSum, id: keyword.id })
+            })
+            .catch(err => new Error(err.message))
+  })
+}
+
+function incrementKeywords(words, amount) {
+  return words.map(word => incrementKeyword(word, amount))
```
Comment (JackyChiu) [inline routes/keyword.js]: You should have the caller do the `.map` and `incrementKeyword` bc it's not clear that calling `incrementKeywords` will return an array or promises

File: test/article.test.js (hunk 23 above)
Comment (kirinrastogi) [inline test/article.test.js]: Its an artifact of my rebase or something. I dont know why this is counting as an addition because I didnt write this

File: routes/article.js (hunk 24 above)
Comment (kirinrastogi) [inline routes/article.js]: Eslint was complaining that err was never used, and I forgot to ask if I should ignore it or do this

File: test/article.test.js (hunk 18 above)
Comment (kirinrastogi) [inline test/article.test.js]: I'll remove the check. If the check is there then mocha doesn't format the error properly (displays it in green text etc)

Comment (JackyChiu) [inline test/article.test.js]: I don't really know when the `err` should be just tbh

File: test/article.test.js (hunk 23 above)
Comment (JackyChiu) [inline test/article.test.js]: Oh odd, is it in a higher scoped `describe` already? If so can you remove this, else add it to higher scope, other tests depend on these values

File: test/article.test.js (hunk 10 above)
Comment (VikramBombhi) [inline test/article.test.js]: kirin y u do this to me

File: test/article.test.js (hunk 6 above)
Comment (VikramBombhi) [inline test/article.test.js]: this is a really good idea. It will definitely improve the security of our code alot

File: test/article.test.js (hunk 19 above)
Comment (VikramBombhi) [inline test/article.test.js]: quick question anyone know if this is the proper way to do this? Just having the id set like this seems like such a hack to me and also seems error prone(what if article gets deleted etc).

File: routes/.eslintrc.json (hunk 12 above)
Comment (VikramBombhi) [inline routes/.eslintrc.json]: He explains why in the pr

File: test/article.test.js (hunk 19 above)
Comment (JackyChiu) [inline test/article.test.js]: Proper way would be to mock a database, test adding an article then taking that id to use for other test

Comment (JackyChiu) [inline test/article.test.js]: Or inject an preexisting article with the id

File: routes/article.js (hunk 24 above)
Comment (JackyChiu) [inline routes/article.js]: Odd Eslint hasn't complained about this to me before, double check or send a pic

File: routes/.eslintrc.json (hunk 12 above)
Comment (JackyChiu) [inline routes/.eslintrc.json]: I think this is a bad idea, for example with this rule we found a better solution to use to use `findByIdAndUpdate`

File: test/article.test.js (hunk 23 above)
Comment (kirinrastogi) [inline test/article.test.js]: I dont think I should change these. The other `lastDate` is a string, this one is a different type, just the same name.

File: routes/article.js (hunk 13 above)
Comment (kirinrastogi) [inline routes/article.js]: Thanks. Is it necessary to runValidators in this case? We're incrementing a Number by either 1 or -1, or not running this, no matter what is posted.

**Comment by JackyChiu** [comment]
@kirinrastogi PR build isn't passing :/

**Comment by kirinrastogi** [comment]
@JackyChiu for some reason the tests sometimes fail remotely, but never locally. fuck my life

**Comment by JackyChiu** [comment]
> @JackyChiu for some reason the tests sometimes fail remotely, but never locally. fuck my life
It passed now LOL, what did you do?

File: routes/.eslintrc.json (hunk 12 above)
Comment (kirinrastogi) [inline routes/.eslintrc.json]: Shouldnt this be moved to a different pr that deals with keywords? I have changed articles to use `findByIdAndUpdate` for now

**Comment by kirinrastogi** [comment]
@JackyChiu I readded .eslintrc.json in routes. But the build failed because of voting test 1

File: routes/.eslintrc.json (hunk 12 above)
Comment (JackyChiu) [inline routes/.eslintrc.json]: @kirinrastogi I strongly disagree with this, let it throw the error now, and when we write better code it won't complain (Moving the logic of creating a new keyword in `fetchArticles`)

Comment (kirinrastogi) [inline routes/.eslintrc.json]: Ok. If we add a `sum` to each keyword then we can use findByIdAndUpdate with $push when updating the keywords too.

Comment (JackyChiu) [inline routes/.eslintrc.json]: Yeah that'll be a lot more clean

File: test/article.test.js (hunk 23 above)
Comment (JackyChiu) [inline test/article.test.js]: When dates get sent as params or queries I'm quite sure they get formatted as strings. So these can be removed

File: routes/article.js [hunk 29]
```diff
@@ -21,46 +21,32 @@ function getAmount(vote) {
...
     return
   }
 
-  let articleChanged = null
-  Article.findById(req.body.id)
-    .then((article) => {
-      article.votes += amount
-      articleChanged = article
-      return article.save()
+  Article.findByIdAndUpdate(
```
Comment (JackyChiu) [inline routes/article.js]: 🥇 💯

File: test/.eslintrc.json [hunk 30]
```diff
@@ -1,5 +1,8 @@
 {
   "rules": {
-    "no-unused-expressions": "off"
+    "no-unused-expressions": "off",
+    "max-len": "off",
+    "no-unused-vars": "off",
+    "no-shadow": "off"
```
Comment (JackyChiu) [inline test/.eslintrc.json]: Why these last two?

File: test/article.test.js [hunk 31]
```diff
@@ -1,17 +1,53 @@
 const chai = require('chai')
 const chaiHttp = require('chai-http')
-const app = require('../server.js')
+const app = require('../server')
 
-chai.should()
+const should = chai.should()
```
Comment (JackyChiu) [inline test/article.test.js]: I'm pretty sure we just need `chai.should()`

File: test/article.test.js [hunk 32]
```diff
@@ -45,25 +87,25 @@ function checkVotesSorted(res) {
...
+  const id = '5846399f9ba87af2501fb035'
+  const fakeId = '5846399f9ba87af2501fb03f'
 
   describe('GET /articles/top', () => {
     it('should have right format response and right properties in articles', (done) => {
       chai.request(app)
         .get('/articles/top')
         .end((err, res) => {
-          checkResponse(res)
+          checkReponse(res)
```
Comment (JackyChiu) [inline test/article.test.js]: I think you forgot to fix the names

File: test/.eslintrc.json (hunk 15 above)
Comment (kirinrastogi) [inline test/.eslintrc.json]: `const should = chai.should()` counts as an unused var in articles.test.js

File: test/.eslintrc.json (hunk 30 above)
Comment (kirinrastogi) [inline test/.eslintrc.json]: Im not done amending my commit. Fixing the next thing you mentioned will get rid of the need for this.

Comment (JackyChiu) [inline test/.eslintrc.json]: Oh okay sounds good 👍

**Comment by kirinrastogi** [comment]
@JackyChiu for the one test that needs the `lastDate` inside an it, should I just rename that var to something else, and use `lastDate: newName` when I query?

File: test/article.test.js (hunk 23 above)
Comment (kirinrastogi) [inline test/article.test.js]: This is so weird. I didn't remove them earlier because the tests would fail, but now when I remove them the tests are passing...

**Comment by JackyChiu** [comment]
> @JackyChiu for the one test that needs the lastDate inside an it, should I just rename that var to something else, and use lastDate: newName when I query?

Don't need to do anything about it, when being sent as a query it became a string anyways

File: test/article.test.js (hunk 23 above)
Comment (JackyChiu) [inline test/article.test.js]: Magic

**Comment by JackyChiu** [comment]
BEAUTIFUL. LOOKS GOOD TO ME 👍 😄 🚢 :shipit:

**Comment by kirinrastogi** [comment]
@VikramBombhi approve changes quickly before travis complains again

**Comment by VikramBombhi** [comment]
ship it!
